From 43b8cf126abd030efc3b720b11c2b153259558ca Mon Sep 17 00:00:00 2001
From: Evan Wang <xswang@marvell.com>
Date: Sat, 14 Jan 2017 00:17:16 +0800
Subject: [PATCH 16/45] phy: u3d: add PHY infrastructure support in USB3 Device
 driver

- The patch added call for PHY init if exists via calling
  APIs (phy_init, phy_power_on, etc.)
- The main purposes to support phy infrastructure are:
  1) Initialize u3d PHY without bootloader dependency
  2) Power on/off in case doing suspend/resume

Change-Id: I79efde4d0a6f0ad2b1b7bca1bd7551a54478a1ae
Signed-off-by: Evan Wang <xswang@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/37074
Tested-by: iSoC Platform CI <ykjenk@marvell.com>
Reviewed-by: Omri Itach <omrii@marvell.com>
---
 .../devicetree/bindings/usb/marvell-u3d.txt   | 32 +++++++++++++++++++
 drivers/usb/gadget/udc/mvebu_u3d.c            | 26 +++++++++++++++
 drivers/usb/gadget/udc/mvebu_u3d.h            |  1 +
 3 files changed, 59 insertions(+)
 create mode 100644 Documentation/devicetree/bindings/usb/marvell-u3d.txt

diff --git a/Documentation/devicetree/bindings/usb/marvell-u3d.txt b/Documentation/devicetree/bindings/usb/marvell-u3d.txt
new file mode 100644
index 000000000000..f7d33dfae105
--- /dev/null
+++ b/Documentation/devicetree/bindings/usb/marvell-u3d.txt
@@ -0,0 +1,32 @@
+Marvell USB3 device controller
+
+This controller is integrated in Armada 38x/3700/8K.
+
+Required properties:
+- compatible	: Must be "marvell,mvebu-u3d"
+- reg		: Physical base address and size of the USB3 device registers map.
+- interrupts	: Should contain single irq line of USB3 device controller
+- clocks	: the clock associated to this controller
+- dma-coherent	: if controller is DMA coherent
+
+Optional properties :
+- vbus-gpio	: VBUS interrupt GPIO pin
+		  set to trigger VBUS event notification to PHY/controller
+		  whether VBUS was powered or not
+- phys		: the COMPHY associated to this controller
+- phy-names:	: the name of COMPHY indicated by above property of phys
+
+Example:
+	u3d@520000 {
+		compatible = "marvell,mvebu-u3d";
+		reg = <0x520000 0x2000>;
+		interrupts = <GIC_SPI 60 IRQ_TYPE_LEVEL_HIGH>;
+		clocks = <&gateclk 24>, <&gateclk 16>;
+		dma-coherent;
+		vbus-gpio = <&gpio1 19 GPIO_ACTIVE_HIGH>;
+		phys = <&a3700_comphy 1 COMPHY_USB3>;
+		phy-names = "usb";
+		status = "disabled";
+	};
+
+
diff --git a/drivers/usb/gadget/udc/mvebu_u3d.c b/drivers/usb/gadget/udc/mvebu_u3d.c
index 9a64928314e4..af62d04d38c7 100644
--- a/drivers/usb/gadget/udc/mvebu_u3d.c
+++ b/drivers/usb/gadget/udc/mvebu_u3d.c
@@ -42,6 +42,7 @@
 #include <linux/highmem.h>
 #include <linux/of_gpio.h>
 #include <linux/gpio.h>
+#include <linux/phy/phy.h>
 
 #include "mvebu_u3d.h"
 
@@ -2398,6 +2399,20 @@ static int mvc2_probe(struct platform_device *pdev)
 		cp->reg->global_control = 0x24;
 	}
 
+	/* Get comphy and init if there is */
+	cp->comphy = devm_of_phy_get(&pdev->dev, pdev->dev.of_node, "usb");
+	if (!IS_ERR(cp->comphy)) {
+		ret = phy_init(cp->comphy);
+		if (ret)
+			goto disable_phy;
+
+		ret = phy_power_on(cp->comphy);
+		if (ret) {
+			phy_exit(cp->comphy);
+			goto disable_phy;
+		}
+	}
+
 	spin_lock_init(&cp->lock);
 
 	/* init irq status */
@@ -2427,6 +2442,11 @@ static int mvc2_probe(struct platform_device *pdev)
 err_qwork:
 	if (cp->qwork)
 		destroy_workqueue(cp->qwork);
+disable_phy:
+	if (cp->comphy) {
+		phy_power_off(cp->comphy);
+		phy_exit(cp->comphy);
+	}
 err_base:
 	iounmap(base);
 err_phybase:
@@ -2464,6 +2484,12 @@ static int mvc2_remove(struct platform_device *dev)
 		destroy_workqueue(cp->qwork);
 	}
 
+	/* PHY exit if there is */
+	if (cp->comphy) {
+		phy_power_off(cp->comphy);
+		phy_exit(cp->comphy);
+	}
+
 	return 0;
 }
 
diff --git a/drivers/usb/gadget/udc/mvebu_u3d.h b/drivers/usb/gadget/udc/mvebu_u3d.h
index b83685def6a2..330125ff2627 100644
--- a/drivers/usb/gadget/udc/mvebu_u3d.h
+++ b/drivers/usb/gadget/udc/mvebu_u3d.h
@@ -350,6 +350,7 @@ struct mvc2 {
 	struct device   *dev;
 	struct clk  *clk;
 	struct usb_phy *phy;
+	struct phy *comphy;
 	int irq;
 	void    __iomem *base;
 	void    __iomem *win_base;
-- 
2.35.3

